mautic/mautic
# Add support email and sms click segment filters and groups "behaviors"

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | enhancement
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This adds support to segment filters for date based clicks from email or text messages. It also adds a filter for click on any link from any SMS (the feature already existed for email but updated the label). 

This PR also updates most if not all of the behavior based filters to make it clearer what the heck they do and places them in their own group in the filters dropdown.

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Send an email with links
2. Click on the links from the email
3. Create a new segment with the "Clicked from an email (date)" filter and select the appropriate operator and timeframe (relative dates are supported like `-90 days`
4. Ensure your contacts are included in the segment
5. Change the filter to be outside the time you clicked and ensure the contact is removed.
6. Repeat with text messages.

Repeat with the `Clicked any link from any email` and `Clicked any link from any text message` filters.

Queries produced: 

For click period:
```
SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT l.id as leadIdPrimary, tUCucvWG.lead_id AS tUCucvWG_lead_id FROM mautic_leads l LEFT JOIN mautic_lead_lists_leads tUCucvWG ON tUCucvWG.lead_id = l.id and (tUCucvWG.leadlist_id = 1) WHERE (l.id IN (SELECT UcMCVNtl.lead_id FROM mautic_page_hits UcMCVNtl WHERE (UcMCVNtl.redirect_id IS NOT NULL) AND (UcMCVNtl.lead_id IS NOT NULL) AND (UcMCVNtl.source = 'email'))) AND (tUCucvWG.lead_id IS NULL)) sss
```

For click based on date/time:
```
SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT l.id as leadIdPrimary, hpflEAgJ.lead_id AS hpflEAgJ_lead_id FROM mautic_leads l LEFT JOIN mautic_lead_lists_leads hpflEAgJ ON hpflEAgJ.lead_id = l.id and (hpflEAgJ.leadlist_id = 1) WHERE (l.id IN (SELECT vImZJfoe.lead_id FROM mautic_page_hits vImZJfoe WHERE (vImZJfoe.redirect_id IS NOT NULL) AND (vImZJfoe.lead_id IS NOT NULL) AND (vImZJfoe.source = 'email') AND (vImZJfoe.date_hit >= '2018-11-24'))) AND (hpflEAgJ.lead_id IS NULL)) sss
```
